app/javascript/packages/memorable-date/index.ts

```diff
@@ -0,0 +1,381 @@
+}
+
+/**
+ * The MemorableDate custom HTML element (WebComponent) provides
+ * a broadly intuitive way for users to enter dates into web applications.
+ *
+ * More about the component here: https://designsystem.digital.gov/components/memorable-date/
+ *
+ * This class facilitates custom error checking and messaging for the MemorableDate
+ * (<lg-memorable-date />) in combination with the ValidatedFieldElement
+ * (<lg-validated-field />). The web server or another source is responsible for
+ * adding the expected child elements for use with this WebComponent.
+ */
+class MemorableDateElement extends HTMLElement {
+  /**
+   * HTML input element for entering the month part of the date
+   */
+  get monthInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__month');
+  }
+
+  /**
+   * HTML input element for entering the day part of the date
+   */
+  get dayInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__day');
+  }
+
+  /**
+   * HTML input element for entering the year part of the date
+   */
+  get yearInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__year');
+  }
+
+  /**
+   * List of HTML input elements for entering month, day and year for a date
+   */
+  get allInputs(): HTMLInputElement[] {
+    const month = this.monthInput;
+    const day = this.dayInput;
+    const year = this.yearInput;
+
+    return [month, day, year].filter(Boolean) as HTMLInputElement[];
+  }
+
+  /**
+   * The configured minimum valid value for the date, based on "min" HTML attribute
+   */
+  get min(): Date | null {
+    return this.getDateAttribute('min');
+  }
+
+  /**
+   * The configured maximum valid value for the date, based on "max" HTML attribute
+   */
+  get max(): Date | null {
+    return this.getDateAttribute('max');
+  }
+
+  connectedCallback() {
+    const { allInputs } = this;
+
+    const inputListener = (e: Event) => {
+      this.validate();
+
+      if (e.type === 'input') {
+        // Artificially trigger input events on all inputs
+        // for input events on one input. This makes the corresponding
+        // <lg-validated-field> elements remove error styling from all
+        // memorable-date fields at the same time as it hides the error
+        // message (instead of only the selected field).
+        const otherInputs = allInputs.filter((i) => i !== e.target);
+        try {
+          otherInputs.forEach((i) => {
+            // Prevent recursion by removing listener temporarily
+            i.removeEventListener('input', inputListener);
+            i.dispatchEvent(new CustomEvent('input', { bubbles: true }));
+          });
+        } finally {
+          otherInputs.forEach((i) => i.addEventListener('input', inputListener));
+        }
+      }
+    };
+
+    allInputs.forEach((i) => {
+      i.addEventListener('input', inputListener);
+      i.addEventListener('invalid', inputListener);
+    });
```

Comment:
Optional: We could rely on event bubbling and add a single event handler to the custom element? (I assume we'd have to make corresponding changes to the temporary removal logic in the event handler).

```suggestion
    this.addEventListener('input', inputListener);
    this.addEventListener('invalid', inputListener);
```

If not, could we at least expand out the single-letter variable for clarity?

```suggestion
    allInputs.forEach((input) => {
      input.addEventListener('input', inputListener);
      input.addEventListener('invalid', inputListener);
    });
```

Reply:
The first suggestion here breaks the following test that I just finished writing:
```
  1) MemorableDateElement
       error message mappings are not empty
         uses customized messages for required validation:

      AssertionError: expected 'Constraints not satisfied' to equal 'Enter a date of birth'
      + expected - actual

      -Constraints not satisfied
      +Enter a date of birth
      
      at expectErrorToEqual (app/javascript/packages/memorable-date/index.spec.ts:60:50)
      at Context.<anonymous> (app/javascript/packages/memorable-date/index.spec.ts:289:7)
      at processImmediate (internal/timers.js:464:21)
```

I renamed the `forEach` parameter for clarity, though.

Reply:
Ah, if I recall correctly, the `invalid` event doesn't bubble, which is probably why it was failing.

I'm fine to keep it as-is, but a couple other ideas:

1. If the idea with the `invalid` event is to try to catch issues when trying to immediately submit a form which hasn't been interacted with, maybe we ought to be calling `this.validate();` in `connectedCallback` to initialize the validity, and then we could get rid of the `invalid` event handler altogether?
2. You can catch the `invalid` event during the capture phase. I tested that this works:
   - `this.addEventListener('invalid', inputListener, true);`

Reply:
I tried the first suggestion, but some of the tests broke because it showed browser built-in validation messages instead of the expected messages. I decided to call `this.validate()` on the `connectedCallback` anyways though since it would present a more up-to-date state for the component.

The second suggestion worked, so I updated to use the `useCapture` flag.

Reply:
> I decided to call `this.validate()` on the `connectedCallback` anyways though since it would present a more up-to-date state for the component.

I was expecting to see this in `connectedCallback`, but I only see it in the event handlers. Were you meaning to call it at the top-level of the `connectedCallback` function? (Which is what I'd expect as well)